src/java.base/share/classes/java/text/StringBuilderBufferProxy.java

```diff
@@ -0,0 +1,188 @@
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package java.text;
+
+/**
+ * Provide the least interfaces that support both:
+ * NumberFormat#format((double|long), StringBuilder, FieldPosition)
+ * and NumberFormat#format(double, StringBuffer, FieldPosition)
+ *
+ */
+interface StringBuilderBufferProxy {
+
+    /**
+     * Return the actual StringBuffer instance
+     *
+     * @return the actual StringBuffer instance
+     */
+    StringBuffer asStringBuffer();
+
+    /**
+     * Return the actual StringBuilder instance.
+     *
+     * @return the actual StringBuilder instance.
+     */
+    StringBuilder asStringBuilder();
+
+    /**
+     * Returns a new {@code String} that contains a subsequence of
+     * characters currently contained in this sequence. The
+     * substring begins at the specified {@code start} and
+     * extends to the character at index {@code end - 1}.
+     *
+     * @param start The beginning index, inclusive.
+     * @param end   The ending index, exclusive.
+     * @return The new string.
+     */
+    String substring(int start, int end);
+
+    /**
+     * Appends the specified string to this character sequence.
+     *
+     * @param str a string.
+     * @return a reference to this object.
+     */
+    StringBuilderBufferProxy append(String str);
+
+    /**
+     * Appends the string representation of the {@code char}
+     * argument to this sequence.
+     *
+     * @param c a {@code char}.
+     * @return a reference to this object.
+     */
+    StringBuilderBufferProxy append(char c);
+
+    /**
+     * Returns the length (character count).
+     *
+     * @return the length of the sequence of characters currently
+     * represented by this object
+     */
+    int length();
+
+    /**
+     * A factory method return an instance with actual implementor
+     * is a StringBuilder instance
+     *
+     * @param sb actual implementor instance
+     * @return an instance that proxy StringBuilder
+     */
+    static StringBuilderBufferProxy of(StringBuilder sb) {
+        return new StringBuilderImpl(sb);
+    }
+
+    /**
+     * A factory method return an instance with actual implementor
+     * is a StringBuffer instance
+     *
+     * @param sb actual implementor instance
+     * @return an instance that proxy StringBuffer
+     */
+    static StringBuilderBufferProxy of(StringBuffer sb) {
+        return new StringBufferImpl(sb);
+    }
+
+    class StringBuilderImpl implements StringBuilderBufferProxy {
```

Comment:
Avoid adding nested classes to interfaces, which are public. These classes are publicly accessible under reflection, and we want to avoid more of such occurrences. See https://bugs.openjdk.org/browse/JDK-8308040?focusedId=14581402&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-14581402

Reply:
Thanks for your suggestion. I split this class into 2 classes: one is the interface, the other is the factory class with package-private.